### ISSUE TITLE
Temporary files left behind when sarracenia is stopped

### ISSUE THREAD

**Issue by andyellisandyellis** [MetPX/sarracenia#591]
Frequently we notice temporary files are being left behind when Sarracenia is stopped. Checking through the logs I can verify that the file was re-downloaded after Sarracenia was restarted and there was no loss of data (at least not in this case). Would it be possible to update Sarracenia to remove any incomplete files when sigterm is received? Alternatively, if there are cases where users want to save incomplete transfers, could a command line option be added to toggle the behaviour?

This problem is occurring as of Sarracenia 2.21.08~ubuntu18.04.1.

Example logs of a transfer that left an incomplete file when Sarracenia was stopped by systemd:
sr_subscribe_ninjo-feed5_01.log.2022-11-15-2022-11-15 18:18:45,325 [INFO] file_log downloaded to: /apps/ninjo/import/radar/in/202211151818~~PPI,PPI_KDP_BALD:URP:CASSF:RADAR:META:20221115181845
sr_subscribe_ninjo-feed5_01.log.2022-11-15-2022-11-15 18:18:45,337 [INFO] Renaming file to: 202211151818~~CLOGZPPI,CLOGZ_BALD:URP:CASFT:RADAR:META:20221115181845
sr_subscribe_ninjo-feed5_01.log.2022-11-15-2022-11-15 18:18:45,381 [INFO] file_log downloaded to: /apps/ninjo/import/radar/in/202211151818~~CLOGZPPI,CLOGZ_BALD:URP:CASFT:RADAR:META:20221115181845
sr_subscribe_ninjo-feed5_01.log.2022-11-15:2022-11-15 18:18:45,382 [INFO] Renaming file to: 202211151818~~PPI,PPI_KDP_BALD:URP:CASFT:RADAR:META:20221115181845
sr_subscribe_ninjo-feed5_01.log.2022-11-15-2022-11-15 18:18:45,417 [INFO] signal stop (SIGTERM)
sr_subscribe_ninjo-feed5_01.log.2022-11-15-2022-11-15 18:18:45,417 [INFO] sr_subscribe stop
sr_subscribe_ninjo-feed5_01.log.2022-11-15-2022-11-15 18:22:51,482 [INFO] sr_subscribe ninjo-feed5 start
--
sr_subscribe_ninjo-feed5_02.log.2022-11-15-2022-11-15 18:22:51,891 [INFO] sr_retry on_heartbeat
sr_subscribe_ninjo-feed5_02.log.2022-11-15-2022-11-15 18:22:51,892 [INFO] No retry in list
sr_subscribe_ninjo-feed5_02.log.2022-11-15-2022-11-15 18:22:51,892 [INFO] sr_retry on_heartbeat elapse 0.001368
sr_subscribe_ninjo-feed5_02.log.2022-11-15:2022-11-15 18:22:51,894 [INFO] Renaming file to: 202211151818~~PPI,PPI_KDP_BALD:URP:CASFT:RADAR:META:20221115182251
sr_subscribe_ninjo-feed5_02.log.2022-11-15:2022-11-15 18:22:51,976 [INFO] file_log downloaded to: /apps/ninjo/import/radar/in/202211151818~~PPI,PPI_KDP_BALD:URP:CASFT:RADAR:META:20221115182251
sr_subscribe_ninjo-feed5_02.log.2022-11-15-2022-11-15 18:22:51,979 [INFO] Renaming file to: 202211151818~~VRPPI,VR_BALD:URP:CASFT:RADAR:META:20221115182251
sr_subscribe_ninjo-feed5_02.log.2022-11-15-2022-11-15 18:22:52,062 [INFO] file_log downloaded to: /apps/ninjo/import/radar/in/202211151818~~VRPPI,VR_BALD:URP:CASFT:RADAR:META:20221115182251
sr_subscribe_ninjo-feed5_02.log.2022-11-15-2022-11-15 18:22:52,075 [INFO] Renaming file to: 202211151818~~VRPPI,VR_LOW:URP:CASET:RADAR:META:20221115182252

**Comment by petersilva** [comment]
fwiw, there was a lot of work done to cleanup shutdowns in sr3.  This is probably already addressed there.  I know that is not helpful for now... just noting it for information's sake.

**Comment by habilinour** [comment]
@petersilva  Yes that the issue that was discussed and reported by @andyellisandyellis

**Comment by habilinour** [comment]
@andyellisandyellis if the file was re-downloaded, the .tmp should have been overwritten ??? unless the filenames were different ??

**Comment by petersilva** [comment]
@habilinour yeah... I don't get that. if it gets downloaded again, the same temp file should be removed and overwritten as part of the later download.

**Comment by andyellisandyellis** [comment]
The timestamp on the end of the file changed so the second file doesn't overwrite the first:

```
sr_subscribe_ninjo-feed5_01.log.2022-11-15:2022-11-15 18:18:45,382 [INFO] Renaming file to: 202211151818~~PPI,PPI_KDP_BALD:URP:CASFT:RADAR:META:2022111518**1845**
sr_subscribe_ninjo-feed5_02.log.2022-11-15:2022-11-15 18:22:51,894 [INFO] Renaming file to: 202211151818~~PPI,PPI_KDP_BALD:URP:CASFT:RADAR:META:2022111518**2251**

```

**Comment by petersilva** [comment]
note to future implementor... what is needed:
  - logic to identify files being transferred ( self.file_inflight )
  - an "abort" function (currently only close exists which would be called in all cases, whether transfer is complete or not.)
  - logic to link signal handler with the abort.
  - likely implementations in v2: sr_util.py, sr_http.py, sr_ftp.py, sr_file.py ... 
  - likely implementation in sr3: transfer/__init__.py , but along the same lines...

**Comment by petersilva** [comment]
I added the "bug" label because that's probably how the end user sees it.  The root cause of this is a custom renamer that causes a inflight place holder files to be named differently on re-send... so it's a defect introduced by someone writing a renamer that breaks what would ordinarily address this issue.

In the current behaviour:
* when a transfer is not completed, the same inflight file (.tmp) is used normally.
  * then the re-send occurs, sarra notices the .tmp file and re-writes it for the new transfer, deleting it when done.  
* in this case, a custom naming scheme is used resulting .tmp file has a different name...
 * when the resend occurs, the old file is not used, so it is not cleaned up.

The optimisation here would be to clean up while shutting down, rather than when re-transfer occurs.  It is likely cleaner to do this, but it requires some work.  It also won't cover cases where abnormal shutdowns occur.
 
I'm wondering... would it not be more straight-forward just not to use a renamer that does this?  You will get the reception time from the file mtime anyways. hmm...

**Comment by petersilva** [comment]
just to understand the reluctance... I'm thinking that in future, especially if the file is big, the sarra logic could see that the tmp file is there, see how much was transferred, and just continue the transfer, rather than starting over... So deleting temporary files will force all restarts to work from scratch, and for big files, that might be a bad thing.

**Comment by andyellisandyellis** [comment]
Yes, that certainly makes sense. Does Sarra have the function to run a checksum of the partially transferred file to verify against the remote side before resuming the transfer?

**Comment by petersilva** [comment]
v2 has partitioned transfers, which is doing checksums on blocks and transferring blocks independently.  We dropped fhe functionality as it was never used in development of sr3, but it was only a matter of prioritization, it will come back once v3 is done.

**Comment by petersilva** [comment]
a good way to address this would be to modify the renamer  plugin to remove the file using an on_stop entry point.
that way the cause is localized with the fix.

**Comment by petersilva** [comment]
partitioned transfers were re-introduced into sr3 in 2024... So now that's also an option... but not used by default.
